### DISCUSSION TITLE
How to reverse proxy websockets?

### DISCUSSION THREAD

**Discussion by tessus** [dani-garcia/vaultwarden#3598] [Q&A]
I am starting this discussion, because there are a few points that are not entirely clear to me. And maybe to others.

I am using the current main branch (at commit 44e9e1a58ed37bf4b352bb499fd3e97adcd3b26b). This code already has the support for websockets on the same port via rocket.
However, I am still using `WEBSOCKET_ENABLED=true` with the following Apache rewrite rules:

```
RewriteEngine On
RewriteCond %{HTTP:Upgrade} =websocket [NC]
RewriteRule /notifications/hub(.*) ws://localhost:3012/$1 [P,L]
ProxyPass / http://localhost:8888/
```

In a PR discussion my question came up how to use the new implementation with `WEBSOCKET_ENABLED=false`.

While the rewrite rule to port 3012 might not be necessary anymore, the protocol still needs to be changed, doesn't it?

The `ProxyPass / http://localhost:8888/` would pass traffic to `http://localhost:8888/`:
```
https://proxy.example.com/notifications/hub?data...  -> http://localhost:8888/notifications/hub?data...
```

But web sockets use a different protocol. So what would happen, if a websocket connection was routed to `http://....`? Shouldn't it be proxied to `ws://....`?

Thoughts?

**Comment by BlackDex** [comment]
Looking at the official documentation, the rewrite rule needs to be changed a bit. https://httpd.apache.org/docs/2.4/mod/mod_proxy_wstunnel.html

```apache
ProxyPass / http://localhost:8888/
RewriteEngine on
RewriteCond %{HTTP:Upgrade} websocket [NC]
RewriteCond %{HTTP:Connection} upgrade [NC]
RewriteRule ^/?(.*) "ws://localhost:8888/$1" [P,L]
```

There are other reverse proxies which handle this then selfs, or need some other small adjustments.

**Reply by BlackDex** [reply]
btw, i actually think this would have been better from the start.

**Reply by tessus** [reply]
Thanks, although there is a slight error in the last line. The pattern is missing. e.g. `^/(.*)` could work.

Just as a fun fact, the `P` flag implies that it is the last rule, so one doesn't need the `L` flag. But it doesn't hurt either.

**Reply by tessus** [reply]
When I get to it, I will run a few tests. No ETA though. I usually either use HAproxy or Apache httpd as a reverse proxy. Traefik I only use in Nomad clusters.

This discussiom should also motivate people to update the [proxy examples](https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples) wiki page. ;-)

Unless somebody beats me to it, I'll update the Apache proxy cfgs (by adding this option and not deleting the previous one) after my tests.

**Reply by BlackDex** [reply]
In the end, i would like to have a clean overview of examples.
Maybe we should move older examples to a lower section on the page. Else it will get very messy and people do not know which version they need to use in the end.

**Comment by stefan0xC** [comment]
> While the rewrite rule to port 3012 might not be necessary anymore, the protocol still needs to be changed, doesn't it?

Nothing should be running on port `3012` with `WEBSOCKET_ENABLED=false`. So the the important part would be to replace the port `3012` with port `8888` in your example (i.e. sending everything to the rocket instance). As far as I understand it the protocol still needs to be changed if you are using `mod_proxy_wstunnel`.

I have not tested this yet but since the websocket URL (where the upgrade happens) is known and if your using [Apache HTTP Server 2.4.47](https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#protoupgrade) (or newer) you might be able to simplify the configuration further using something like the following `mod_proxy` directive:
```apache
ProxyPass /notifications http://localhost:8888/notifications upgrade=websocket
ProxyPass / http://localhost:8888/
```

**Reply by BlackDex** [reply]
Well, the reason is, that there could be more entry-points for web-sockets like #3592. Because of that, it's better to not use a specific path, and just detect if there is a web-socket request sent and upgrade when needed, like is done in my example.

**Reply by stefan0xC** [reply]
Ah, I have not seen that. Can't we assume that every endpoint under the `/notifications` path might upgrade to a websocket connection?

**Reply by BlackDex** [reply]
Why should we? I would rather have a more general / standard config then a fully customized for Vaultwarden config.

**Reply by stefan0xC** [reply]
Because it seems like a safe assumption given that there is currently only one path (and there will probably only be two paths) to which this applies? I think that the assumption that every path might upgrade to a websocket connection is a bit overkill (even if the overhead of checking for those headers is probably negligible).

**Reply by BlackDex** [reply]
The problem is, if there are new ways or endpoints add, people need to change there reverse proxy config again. It's better to have a general working example.

**Reply by tessus** [reply]
stefan0xC was already on the right track.

As of Apache HTTP Server 2.4.47, the following is all that is needed to upgrade websocket connections. No more rewrite condition and rules.

```
ProxyPass / http://localhost:8888/ upgrade=websocket
```

There is no need to specify a `/notification` path. Upgrade only takes place when requested
AND accepted/switched by the backend server.

I'll add another entry below for Apache. With time maybe people add the setups for other reverse proxies as well. In which case at one point we just have to copy and paste that info to the wiki. ;-)

**Comment by ShlomiD83** [comment]
I don't know if it's related to the discussion about WS but I think the chrome extension doesn't auto update when making changes (e.g. adding a folder), the web vault and android app update almost instantly.
that wasn't the case a few weeks ago I believe, it started after the push notifications were implemented.
if it's not related please guide me if I should open an issue or try somewhere else.

**Reply by BlackDex** [reply]
It seems to work for me. Could be an extension issue.
I would suggest to try a clean profile, and also what happens with an other browser to see what that does.

**Reply by ShlomiD83** [reply]
I tried with chrome in incognito, didn't work.
Installed Firefox, still doesn't work.

WS notifications work, I get a 101 when checking in developer mode, and the web vault works fine as well.
I'm using the testing-alpine image if that matters, tested the Debian based testing image and the same problem.
what else should I try?

**Comment by ShlomiD83** [comment]
Reverting to the main branch vaultwarden/server:alpine image solves the issue.
please advise.

**Reply by BlackDex** [reply]
Please create a new discussion regarding this and provide the support string which you can generate via the admin diagnostics page.

**Reply by ShlomiD83** [reply]
No problem.

**Comment by ShlomiD83** [comment]
So what is the recommended way of using websockets?
Remove the env variable, the port mapping and the Nginx proxy manager config for /notifications/hub?

**Comment by tessus** [comment]
# Apache httpd

In vaultwarden set `WEBSOCKET_ENABLED=false`

## Apache HTTP Server >= 2.4.47

```apache
ProxyPass / http://localhost:8888/ upgrade=websocket
```

## Apache HTTP Server < 2.4.47

```apache
ProxyPass / http://localhost:8888/
RewriteEngine   on
RewriteCond     %{HTTP:Upgrade}       websocket                   [NC]
RewriteCond     %{HTTP:Connection}    upgrade                     [NC]
RewriteRule     ^/?(.*)               "ws://localhost:8888/$1"    [P,L]
```

In case your distro patched Apache HTTP Server < 2.4.47 with the mitigation for [CVE-2024-38474](https://www.cve.org/CVERecord?id=CVE-2024-38474), the rewrite rule must be changed to:

```apache
RewriteRule     ^/?(.*)               "ws://localhost:8888/$1"    [P,L,UnsafeAllow3F]
```

_(Thanks to @roburb for bringing this to my attention.)_

**Reply by roburb** [reply]
**Apache 2.4.37 Alamalinux 8.10**
I have installed Vaultwarden in sublocation and used example on wiki page
https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples (by [@agentdr8](https://github.com/agentdr8))

I have to add **"UnsafeAllow3F"** => all works
...
_RewriteRule /notifications/hub(.*) ws://127.0.0.1:8070/vaultwarden/notifications/hub/$1 [P,L,UnsafeAllow3F]_
...
maybe this can help someone :-)

**Reply by tessus** [reply]
@roburb This makes no sense. Apache 2.4.37 does not have the flag `UnsafeAllow3F` in mod_rewrite.

P.S.: I checked the source code of 2.4.37 and the flag is not there. Only after Apache 2.6.61 a rule could result in an error, if `UnsafeAllow3F` was not set but worked before 2.6.61.

I also used the above ruleset without the `UnsafeAllow3F` flag, before I switched to the one-liner and it worked as it should.

**Reply by roburb** [reply]
I don't know, it is my experience.
Without the flag it doesn't work, with the flag it works.

_# httpd -v
Server version: Apache/2.4.37 (AlmaLinux)
Server built:   Aug 12 2024 02:30:19_

As I said "Maybe it could help" 
That is all :-)

**Reply by tessus** [reply]
Which brings me back to me initial sentiment: it makes no sense. How can a flag help, if it does not exist in that version?

**Reply by BlackDex** [reply]
@tessus then why is it mentioned in the docs? https://httpd.apache.org/docs/2.4/rewrite/flags.html#flag_unsafe_allow_3f

**Reply by tessus** [reply]
@BlackDex I do not know. Maybe they forgot to add the info at which version it is available.

You can check `modules/mappers/mod_rewrite.c`. These are the flags that are defined (in 2.4.37):

```c
#define RULEFLAG_NONE               (1<<0)
#define RULEFLAG_FORCEREDIRECT      (1<<1)
#define RULEFLAG_LASTRULE           (1<<2)
#define RULEFLAG_NEWROUND           (1<<3)
#define RULEFLAG_CHAIN              (1<<4)
#define RULEFLAG_IGNOREONSUBREQ     (1<<5)
#define RULEFLAG_NOTMATCH           (1<<6)
#define RULEFLAG_PROXY              (1<<7)
#define RULEFLAG_PASSTHROUGH        (1<<8)
#define RULEFLAG_QSAPPEND           (1<<9)
#define RULEFLAG_NOCASE             (1<<10)
#define RULEFLAG_NOESCAPE           (1<<11)
#define RULEFLAG_NOSUB              (1<<12)
#define RULEFLAG_STATUS             (1<<13)
#define RULEFLAG_ESCAPEBACKREF      (1<<14)
#define RULEFLAG_DISCARDPATHINFO    (1<<15)
#define RULEFLAG_QSDISCARD          (1<<16)
#define RULEFLAG_END                (1<<17)
#define RULEFLAG_ESCAPENOPLUS       (1<<18)
#define RULEFLAG_QSLAST             (1<<19)
```

**Reply by BlackDex** [reply]
As it looks like there has been a CVE Report regarding this https://www.cve.org/CVERecord?id=CVE-2024-38474, i think AlmaLinux patched the version they use instead of updating to 2.4.60 (Which has this built-in), which happens all the time with other distro's too. They just release a custom patched version without changing the version that much.

Some security tools then trip over the version number not knowing the server is already using a patched version.

**Reply by tessus** [reply]
I see, now this make sense and it would also explain why it fails on Alma Linux without that flag.

**Comment by netboy3** [comment]
How would those rules work with a non-root domain setup, I.E. https://xxx.yyy.zzz/bitwarden/....
My environment is set with `DOMAIN=https://xxx.yyy.zzz/bitwarden`, but vaultwarden seems to barf with 404's on any requests coming for ".../anonymous-hub/?"

**Reply by BlackDex** [reply]
The exact same as above as far as i know.

**Reply by tessus** [reply]
Since you did not tell us which version of Apache, I will only write down the directive for >= 2.4.47.

```apache
ProxyPass /bitwarden/ http://localhost:8888/ upgrade=websocket
```

**Reply by netboy3** [reply]
Apache 2.4.6

**Reply by tessus** [reply]
```apache
ProxyPass /bitwarden/ http://localhost:8888/
RewriteEngine   on
RewriteCond     %{HTTP:Upgrade}       websocket                   [NC]
RewriteCond     %{HTTP:Connection}    upgrade                     [NC]
RewriteRule     ^/bitwarden/?(.*)     "ws://localhost:8888/$1"    [P,L]
```

**Reply by netboy3** [reply]
Still not working. This is what I see in the extended log:
```
[2023-09-29 14:43:47.322][request][INFO] GET /notifications/anonymous-hub?Token=5ab227c7-dbd7-4252-8fc1-
[2023-09-29 14:43:47.322][rocket::server::_][ERROR] No matching routes for GET /notifications/anonymous-hub?Token=5ab227c7-dbd7-4252-8fc1-88ebe1beffef.
[2023-09-29 14:43:47.322][rocket::server::_][WARN] No 404 catcher registered. Using Rocket default.
[2023-09-29 14:43:47.322][response][INFO] 404 Not Found
```

**Reply by netboy3** [reply]
What should the `GET` request look like when it hits the websocket? I have a pretty good understanding of mod_rewrite and can experiment with this, but I can't really figure out what is expected there and whether vaultwarden's `DOMAIN` attribute makes any modifications to whatever is expected there.

**Reply by stefan0xC** [reply]
What version of vaultwarden are you using? (That endpoint was only added in the latest release,`1.29.2`)

Also can you try adding a trailing slash to `DOMAIN=https://xxx.yyy.zzz/bitwarden/` as [recommended by the proxy examples](https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples)?

**Reply by netboy3** [reply]
> What version of vaultwarden are you using? (That endpoint was only added in the latest release,1.29.2)

1.29.2-bc26bfa5

> Also can you try adding a trailing slash to DOMAIN=https://xxx.yyy.zzz/bitwarden/ as [recommended by the proxy examples](https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples)?

Didn't fix the issue. Still getting no matching routes

This seems to be a bug with how `DOMAIN` effects the endpoints. To test this out I've setup another virtual Apache server and pointed it at vaultwarden, then disabled `DOMAIN` and it worked perfectly. The log shows that in both cases the `GET` request has the same path ("GET /notifications/anonymous-hub?Token=..."). Once `DOMAIN` is configured, I'm getting no matching routes. This also effects "/notifications/hub?..." which also fails if `DOMAIN` is configured.

**Reply by BlackDex** [reply]
If you use a sub-path, you also need to input that into the domain variable as is described in the wiki in several places.
So `https://secure.domain.tld/vaultwarden` also, check if you do not have a config.json file which overrules the env settings.

**Reply by netboy3** [reply]
I do not have config.json and never setup anything in the Admin UI. When I mention configuring the `DOMAIN` variable it means assigning it the proper value ("DOMAIN=https://xxx.yyy.zzz/bitwarden/"). Setting it up seems to be the root cause of this problem. Was the sub-path feature tested with the new, unified port setup?

**Reply by BlackDex** [reply]
It is tested yes, also, it isn't any different since the logic is the same for all paths. 

But i think it should be the following to get it working.
Bit that is an Apache item, and not Vaultwarden.

```apache
RewriteRule     ^/bitwarden/?(.*)     "ws://localhost:8888/bitwarden/$1"    [P,L]
```

**Reply by netboy3** [reply]
@BlackDex thank you! Your rule made it work. It was stupid of me not to consider prefixing the endpoint with the sub-path too. Might be a good thing to also update the WiKi as the current example does not do that.

**Reply by tessus** [reply]
@BlackDex @netboy3 this is rather confusing to me. Usually the backend does not use a subpath, which is why there are reverse proxy directives. In certain cases I have seen config options to also use a subpath at the backend, but I haven't seen one for vaultwarden.

Tbh, I do not understand how that RewriteRule can work.

**Reply by BlackDex** [reply]
Well, Vaultwarden needs too know the whole URI. Else it can't produce the right links or Webauthn config to be valid. It is also used by the admin to make sure the links are going to the correct location, and to validate the URI/Domain is correct.

Three reason this works, is because you ProxyPass to the `/`, and that seems to ommit the sub-path for the rewrite rule, because you also match on that specific path, it does pass that path on to all the others.

Since Vaultwarden adjusts al the paths to use the sub-path everything without that path will be invalid.

**Reply by netboy3** [reply]
Also, the RewriteRule actually only effects the WS handshake phase which is basically HTTP. It's being used (as BlackDex described) to let vaultwarden know what endpoint needs to be accessed. It seems that once you configure the `DOMAIN` attribute, the whole vaultwarden namespace will be prefixed with the subpath. So it's critical to form the endpoint handshake URLs such as "/<optional_subpath>/notifications/hub?...". Once a request comes from the bitwarden client, the URL will be adjusted. Here's the request from the bitwarden browser extension hitting vaultwarden after the URL mangling:

```
GET /bitwarden/notifications/hub?access_token=XXX
HTTP/1.1
Host: xxx.yyy.zzz
Pragma: no-cache
Cache-Control: no-cache
User-Agent: XXX
Origin: https://xxx.yyy.zzz
Sec-WebSocket-Version: 13
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US
Sec-WebSocket-Key: XXX
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
X-Real-IP: 10.10.10.10
X-Forwarded-For: 10.10.10.10
X-Forwarded-Host: xxx.yyy.zzz
X-Forwarded-Server: xxx.yyy.zzz
Upgrade: WebSocket
Connection: Upgrade
```

Because the `GET` request hit a known endpoint, and has the upgrade header, it will trigger Rocket to switch protocol to WebSocket. Before switching, it will send an approval HTTP response to the client:

```
HTTP/1.1 101 Switching Protocols
sec-websocket-version: 13
sec-websocket-accept: XXX
server: Rocket
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
permissions-policy: XXX
referrer-policy: same-origin
x-xss-protection: 0
content-security-policy: XXX
cache-control: no-cache, no-store, max-age=0
access-control-allow-origin: https://xxx.yyy.zzz
connection: Upgrade
upgrade: websocket
content-length: 0
date: Sat, 30 Sep 2023 13:47:48 GMT
```

Apache (or any other proxy) sits in the middle of this traffic and is monitoring it. Once it sees a successful handshake such as above it will create a tunnel between the client and the backend server to exchange WS messages. The next packet sent out from vaultwarden is a WS ping and is simply forwarded by the Apache tunnel to the client which responds with a WS pong. This now goes on. When a WS tunnel is already active there's obviously no meaning to path.

**Reply by tessus** [reply]
My point is that 

```apache
RewriteRule     ^/bitwarden/?(.*)     "ws://localhost:8888/$1"    [P,L]
```

reverse-proxies `https://example.com/bitwarden/notifications/hub?access_token=XXX` to `ws://localhost:8888/notifications/hub?access_token=XXX`. And this is IMO the correct way, because vaultwarden does not have a config option to specify a subpath.

The only config option that on can set in vaultwarden is the **external** request URI, which is `https://example.com/bitwarden`

So, if you use the directive 

```apache
RewriteRule     ^/bitwarden/?(.*)     "ws://localhost:8888/bitwarden/$1"    [P,L]
```

a request will be reverse-proxied `ws://localhost:8888/bitwarden/notifications/hub?access_token=XXX`, but vaultwarden should not have an endpoint: `ws://localhost:8888/bitwarden/`, because `bitwarden` was never specified in the vaultwarden config. Except for the EXTERNAL request URI, which is not the backend...

**Reply by BlackDex** [reply]
Vaultwarden **does** have a setting for the sub-path, that needs to be done in the `DOMAIN` config, as described on the wiki.

**Reply by netboy3** [reply]
> because vaultwarden does not have a config option to specify a subpath

But it does. That's what the `DOMAIN` variable does. It prefixes the whole Rocket namespace with the subpath specified there.

**Reply by tessus** [reply]
This is exactly my point. The `DOMAIN` variable is the external request URI. vaultwarden does not listen on the external URI. The reverse proxy does. vaultarden listens on `ROCKET_ADDRESS` and `ROCKET_PORT`. There is no variable for `ROCKET_PATH`.

Unless the vaultwarden server uses the external URI to build a local URI for the app server. Which is not how that usually works.

Look at any other server that can be reverse-proxied. There are usually 2 variables: one for the external request URI and one for the local app server path. e.g. they could be different.

**Reply by BlackDex** [reply]
Again, we use the `DOMAIN` variable. It needs to match. We extract the path from that variable and prepend that to the endpoints.

**Reply by netboy3** [reply]
Here's how it's done in the code:
First, the path is extracted from the DOMAIN URL and assigned to `domain_path` (settings struct member):
https://github.com/dani-garcia/vaultwarden/blob/bc26bfa589c007da9b9be37e1172060f38a948b9/src/config.rs#L428

Then `domain_path` is being assigned to a `basepath` local variable:
https://github.com/dani-garcia/vaultwarden/blob/bc26bfa589c007da9b9be37e1172060f38a948b9/src/main.rs#L514

Which in turn is used as a prefix when mounting/registering routes/catchers during Rocket launch:
https://github.com/dani-garcia/vaultwarden/blob/bc26bfa589c007da9b9be37e1172060f38a948b9/src/main.rs#L527-L536
